diff --git a/i18n_menu/i18n_menu.module b/i18n_menu/i18n_menu.module
index ab85792..d59c8c1 100644
--- a/i18n_menu/i18n_menu.module
+++ b/i18n_menu/i18n_menu.module
@@ -515,6 +515,8 @@ function _i18n_menu_link_description($link, $langcode = NULL) {
  * Check whether this link is to be processed by i18n_menu and start processing.
  */
 function _i18n_menu_link_process(&$link) {
+  global $language;
+
   // Only visible links that have a language property and haven't been processed
   // before. We also check that they belong to a menu with language options.
   if (empty($link['i18n_menu']) && !empty($link['language']) && !empty($link['access']) && empty($link['hidden']) && i18n_menu_mode($link['menu_name'])) {
@@ -530,7 +532,7 @@ function _i18n_menu_link_process(&$link) {
       }
     }
     // Skip if administering this menu item through the node edit form.
-    elseif (arg(0) == 'node' && arg(2) == 'edit') {
+    elseif (arg(0) == 'node' && arg(2) == 'edit' && $link['link_path'] == arg(0) . '/' . arg(1) && $link['language'] == $language->language) {
       return FALSE;
     }
     return TRUE;
